### PR TITLE
Don't load entire file into memory during uploads over SSH

### PR DESCRIPTION
**Pull request by y0ssar1an**
The `ssh` package uses `scp` to upload files to a VM. The `scp` protocol requires that a header be printed which contains the size of the file to be transferred. Our previous solution to this problem was to load the entire file into memory and check `len(file)`. This is obviously a problem if you're trying to upload large files.

The new code checks to see if the `src io.Reader` is an `*os.File`. If so, it gets the file size with a `Stat()` call, without consuming the reader. This solves the memory usage problem for files only. Fortunately, that's probably 99.9% of the actual usage of `ssh.Upload`.

P.S. This is branched off the dep update in #135. It should only be merged after #135.

### CODE REVIEW

**Comment by tcahill** [comment]
Did you mean to include the dependency update in this PR? Seems like it should be separate.

File: ssh/ssh.go [hunk 1]
```diff
@@ -352,6 +353,25 @@ func (client *SSHClient) Upload(src io.Reader, dst string, mode uint32) error {
...
+func getFileSize(src io.Reader) (r io.Reader, fileSize int, err error) {
+	if f, ok := src.(*os.File); ok {
+		fi, err := f.Stat()
+		if err != nil {
+			return nil, 0, err
+		}
+		return src, int(fi.Size()), nil
+	}
+
+	fileContent, err := ioutil.ReadAll(src)
```
Comment (levb) [inline ssh/ssh.go]: I don't think we should *ever* use ioutil, specifically `ReadAll()`
Probably better throw an error here... or require that the length is passed in as a parameter

**Comment by y0ssar1an** [comment]
Yeah, I was hoping for a quick merge of the dep update, since it's just a housekeeping update. There's a low risk of breakages since it's effectively already been tested by dep updates to `apcera-setup`.

**Comment by y0ssar1an** [comment]
@levb makes a good point. The current code solves the memory usage problem for files only. We could solve the problem for all `io.Reader`s by requiring that the caller to pass the size of the object to be uploaded. But that's a breaking change to the package API.

So our choices are:
1) Make a breaking change to `ssh.Upload`, but fix the problem The Right Way:tm:
2) Avoid the breaking change, but settle for a so-so fix.

**Comment by mbhinder** [comment]
We could also add a new method and deprecate the old one while adding the code to avoid the `ReadAll` if the passed in `Reader` is a file.

**Comment by y0ssar1an** [comment]
UPDATE: I changed the code to return an error if the `io.Reader` being uploaded is not a file. This let's us remove the `ioutil.ReadAll` command without changing the `Upload` signature (by adding a `fileSize` argument.

**Pull request merged**